x/oracle/keeper/prophecy.go

```diff
@@ -0,0 +1,135 @@
+
+	store.Set(storePrefix, k.cdc.MustMarshalBinaryBare(&prophecy))
+}
+
+// GetProphecyInfo return a prophecy's signatures
+func (k Keeper) GetProphecyInfo(ctx sdk.Context, prophecyID []byte) (types.ProphecyInfo, bool) {
+	var prophecySignatures types.ProphecyInfo
+	store := ctx.KVStore(k.storeKey)
+
+	// storePrefix := append(types.SignaturePrefix, prophecyID[:]...)
+	bz := store.Get(append(types.SignaturePrefix, prophecyID[:]...))
+	if bz == nil {
+		return types.ProphecyInfo{}, false
+	}
+
+	k.cdc.MustUnmarshalBinaryBare(bz, &prophecySignatures)
+
+	return prophecySignatures, true
+}
+
+// SetProphecyInfo saves a prophecy with an initial value
+func (k Keeper) SetProphecyInfo(ctx sdk.Context, prophecyID []byte, networkDescriptor types.NetworkDescriptor,
+	cosmosSender string,
+	cosmosSenderSequence uint64,
+	ethereumReceiver string,
+	tokenSymbol string,
+	tokenAmount sdk.Int,
+	crosschainFee sdk.Int,
+	doublePeg bool,
+	globalNonce uint64) error {
+
+	store := ctx.KVStore(k.storeKey)
+
+	storePrefix := append(types.SignaturePrefix, prophecyID[:]...)
+
+	prophecyInfo := types.ProphecyInfo{
+		ProphecyId:           prophecyID,
+		NetworkDescriptor:    networkDescriptor,
+		CosmosSender:         cosmosSender,
+		CosmosSenderSequence: cosmosSenderSequence,
+		EthereumReceiver:     ethereumReceiver,
+		TokenSymbol:          tokenSymbol,
+		TokenAmount:          tokenAmount,
+		DoublePeg:            doublePeg,
+		GlobalNonce:          globalNonce,
+		CrosschainFee:        crosschainFee,
+		EthereumAddress:      []string{},
+		Signatures:           []string{},
+		BlockNumber:          uint64(k.currentHeight),
+	}
+
+	store.Set(storePrefix, k.cdc.MustMarshalBinaryBare(&prophecyInfo))
+	return nil
+}
+
+// AppendSignature add a new ethereum address and signature to prophecy
+func (k Keeper) AppendSignature(ctx sdk.Context, prophecyID []byte, ethereumAddress, signature string) error {
+	store := ctx.KVStore(k.storeKey)
+
+	prophecySignatures, ok := k.GetProphecyInfo(ctx, prophecyID)
+	if !ok {
+		return errors.New("can not get the prophecy")
+	}
+
+	prophecySignatures.EthereumAddress = append(prophecySignatures.EthereumAddress, ethereumAddress)
+	prophecySignatures.Signatures = append(prophecySignatures.Signatures, signature)
+
+	storePrefix := append(types.SignaturePrefix, prophecyID[:]...)
+
+	store.Set(storePrefix, k.cdc.MustMarshalBinaryBare(&prophecySignatures))
+	return nil
+}
+
+// CleanUpProphecy clean up outdated prophecy
+func (k Keeper) CleanUpProphecy(ctx sdk.Context) {
+	var prophecyInfo types.ProphecyInfo
+	currentHeight := uint64(k.currentHeight)
+
+	store := ctx.KVStore(k.storeKey)
+	iter := sdk.KVStorePrefixIterator(store, types.SignaturePrefix)
+	for ; iter.Valid(); iter.Next() {
+		k.cdc.MustUnmarshalBinaryBare(iter.Value(), &prophecyInfo)
+		if prophecyInfo.BlockNumber-currentHeight > ProphecyLiftTime {
+			storePrefix := append(types.SignaturePrefix, prophecyInfo.ProphecyId[:]...)
+			store.Delete(storePrefix)
```

Comment:
Here if we end up deleting, we should update the k.currentHeight to be the current block so that we do not continually scan every block after the first 1000 blocks.

Reply:
the currentHeight is keeping updated via following method for every new block
func (am AppModule) BeginBlock(ctx sdk.Context, block abci.RequestBeginBlock) {
	am.keeper.UpdateCurrentHeight(block.Header.Height)